src/stream/src/executor/temporal_join.rs

```diff
@@ -310,6 +309,170 @@ async fn align_input(left: Executor, right: Executor) {
     }
 }
 
+mod phase1 {
+    use futures_async_stream::try_stream;
+    use risingwave_common::array::stream_chunk_builder::StreamChunkBuilder;
+    use risingwave_common::array::{Op, StreamChunk};
+    use risingwave_common::hash::{HashKey, NullBitmap};
+    use risingwave_common::row::{self, Row, RowExt};
+    use risingwave_common::types::{DataType, DatumRef};
+    use risingwave_common::util::iter_util::ZipEqDebug;
+    use risingwave_hummock_sdk::HummockEpoch;
+    use risingwave_storage::StateStore;
+
+    use super::{StreamExecutorError, TemporalSide};
+
+    pub(super) trait Phase1Evaluation {
+        /// Called when a matched row is found.
+        #[must_use = "consume chunk if produced"]
+        fn append_matched_row(
+            op: Op,
+            builder: &mut StreamChunkBuilder,
+            left_row: impl Row,
+            right_row: impl Row,
+        ) -> Option<StreamChunk>;
+
+        /// Called when all matched rows of a join key are appended.
+        #[must_use = "consume chunk if produced"]
+        fn match_end(
+            builder: &mut StreamChunkBuilder,
+            op: Op,
+            left_row: impl Row,
+            right_size: usize,
+            matched: bool,
+        ) -> Option<StreamChunk>;
+    }
+
+    pub(super) struct Inner;
+    pub(super) struct LeftOuter;
+    pub(super) struct LeftOuterWithCond;
```

Comment:
😄 Looks really similar to https://github.com/risingwavelabs/risingwave/blob/8ea55a038514b4be7bdd4cc97c49016b5eb4b564/src/stream/src/executor/aggregation/agg_group.rs#L36-L51

I slightly prefer naming the structs by their immediate semantics instead of the join modes. However it's not a strong opinion, up to you😄